### ISSUE TITLE
[Question] Reproducible QwenFAST training configuration and malformed action decoding for RoboTwin 2.0

### ISSUE THREAD

**Issue by chaojupup** [starVLA/starVLA#428]
Hi StarVLA team,

Thank you for releasing StarVLA.

We are training QwenFAST on the RoboTwin Clean 50-task dataset. We would like to confirm the recommended RoboTwin configuration and understand how malformed FAST sequences should be handled during closed-loop evaluation.

## Environment

* StarVLA branch: `starVLA_dev`
* StarVLA base commit: `c00b8e18c1b5ba6228fd68a8f38acca8df4e938a`
* Python: `3.10.20`
* PyTorch: `2.6.0+cu124`
* Transformers: `5.9.0`
* Tokenizers: `0.22.2`
* Attention implementation: `sdpa`
* Framework: `QwenFast`
* VLM: Qwen3.5-4B with 2,048 FAST action special tokens
* FAST tokenizer: `physical-intelligence/fast`
* Dataset: RoboTwin Clean, 50 tasks
* State input: disabled
* Fine-tuning: full fine-tuning
* GPUs: 4
* Per-device batch size: 4
* Global batch size: 16
* Gradient accumulation steps: 1
* Checkpoint: 90K steps

The repository contains local adaptations for Qwen3.5, RoboTwin Clean data registration, and strict FAST sequence validation on top of the base commit listed above.

## Action configuration

* Action type: absolute joint targets
* Dataset action type: `abs_qpos`
* Action mode: `abs`
* Action dimension: 14

  * Left arm: 6 joints
  * Left gripper: 1
  * Right arm: 6 joints
  * Right gripper: 1

* Action horizon: 50

We selected `action_dim=14` and `action_horizon=50` based on the official RoboTwin absolute-action configuration:

```text
examples/simBenchmarks/Robotwin/train_files/starvla_cotrain_robotwin_abs.yaml
```

However, this configuration uses QwenOFT rather than QwenFAST, so we are unsure whether the same action horizon is recommended for the autoregressive FAST action representation.

## Generation configuration

We currently use:

```python
do_sample = False
enable_thinking = False
eos_token_id = 248046       # <|im_end|>
pad_token_id = 248046       # generation uses EOS as padding
max_new_tokens = 1024
```

The tokenizer's native padding token is:

```python
tokenizer.pad_token_id = 248044  # <|endoftext|>
```

but the generation call explicitly uses `pad_token_id=eos_token_id`.

Please let us know whether QwenFAST requires any additional generation constraints or stopping criteria.

## Observed problem

For an action chunk with shape `(50, 14)`, FAST BPE decompression should produce exactly:

```text
50 × 14 = 700 DCT coefficients
```

We checked the training data and labels:

* Raw action shape is always `(50, 14)`.
* FAST encode/decode round trips are valid.
* The decompressed target length is always 700.
* On 10 initial observations sampled from the training distribution, the 90K checkpoint produced 10 out of 10 valid FAST sequences.

However, malformed sequences appear during closed-loop RoboTwin evaluation after the robot begins moving.

### `adjust_bottle`

* The first four generated action chunks were valid.
* Later generations repeatedly decompressed to 693 coefficients.
* Expected length: 700.

### `turn_switch`

* One generation decompressed to 711 coefficients.
* Later generations produced 692, 693, or 699 coefficients.
* Expected length: 700.

Because greedy generation is deterministic, retrying with the same observation produces the same malformed sequence.

We currently reject malformed sequences instead of executing an all-zero fallback or applying padding or truncation.

## Qwen3.5 template consistency

We also found that the Qwen3.5 chat and thinking templates must remain consistent between training and inference.

We use:

```python
enable_thinking = False
```

and explicitly set the Qwen3.5 EOS token.

Before this change, the inference prompt ended in a different thinking context from the training prompt, causing widespread malformed FAST generations. Disabling thinking consistently fixed that mismatch, but occasional malformed FAST sequences still occur during closed-loop rollout.

## Questions

1. Is there an officially validated QwenFAST configuration or YAML file for RoboTwin?

2. Is `action_horizon=50` expected to work with QwenFAST for 14-dimensional absolute actions, or would you recommend a shorter horizon such as 8 or 16?

3. How does the official StarVLA evaluation handle generated FAST sequences whose decompressed coefficient count does not equal:

   ```text
   action_horizon × action_dim
   ```

4. In this situation, would you recommend:

   * rejecting the prediction,
   * executing an all-zero action,
   * padding or truncating the coefficients,
   * regenerating with sampling,
   * or generating multiple candidates and selecting a valid sequence?

5. Have you observed similar malformed FAST sequences with the officially tested Qwen2.5-VL or Qwen3-VL QwenFAST models?

6. Has Qwen3.5 been internally tested with QwenFAST, or are there additional tokenizer, chat-template, EOS, or stopping-criterion changes required compared with Qwen3-VL?

The official LIBERO QwenFAST configuration uses:

```text
action_dim = 7
action_horizon = 8
```

which corresponds to only 56 decompressed coefficients. Our RoboTwin configuration requires 700 coefficients. We are wondering whether the substantially longer autoregressive FAST sequence is the main cause of the malformed outputs.

A minimal QwenFAST RoboTwin YAML configuration, generation configuration, or patch based on the existing QwenOFT RoboTwin configuration would be very helpful.

Thank you!

**Comment by JinhuiYE** [comment]
Thanks for your question. FAST tokenization is based on frequency‑domain analysis (DCT), so the token length is not directly determined by action_horizon * action_dim – those are only used to reshape the decoded latent back into the action matrix.

However, the decoding stage involves a strict discrete‑to‑continuous mapping (similar to BPE, where discrete tokens are converted back to continuous frequency coefficients). If the generated token sequence cannot be reshaped into the expected matrix shape – for example, due to wrong length or invalid tokens – the reconstruction fails. Moreover, the entire generated token sequence must satisfy rigid constraints on length and format; any deviation will produce a malformed output.

Therefore, the malformed sequence you encountered is almost certainly due to a configuration mismatch between training and inference (e.g., chat template with enable_thinking, eos_token_id, pad_token_id, or sampling parameters). Please verify that all these settings are identical. You may add a fallback (e.g., a safe default action) for temporary handling, but the root cause lies in the configuration inconsistency.

**Comment by JinhuiYE** [comment]
Follow-up clarification: the public repository does not ship an officially validated QwenFAST RoboTwin or Qwen3.5 recipe, so we cannot claim that horizon 50 is validated or that a configuration mismatch is the confirmed root cause. The QwenOFT horizon-50 configuration does not validate the materially different autoregressive FAST regime. Malformed decoded action sequences should be rejected, never silently padded or truncated; a deployment path should stop or hold safely, or replan. Since training and inference template consistency has already been checked, diagnosing any remaining code bug requires the exact generation trace, configuration, and checkpoint. We are closing this as answered regarding the current public support boundary; please reopen with a minimal reproducer if the mismatch remains.
